### PR TITLE
add lbName annotation

### PR DESCRIPTION
**Pull request by agadelshin**
this PR create feature for #369 

I've created annotation "kubernetes.io/ingress.loadbalancer-name" to share LB between ingresses. Work still in progress, but it contains a lot of changes, so please look at this PR to control progress.

### CODE REVIEW

**Comment by pondohva** [comment]
I still need to do smth with TLS configuration, but simple case with N ingresses with the same LB name in annotation works.

**Comment by pondohva** [comment]
@rramkumar1 are you OK with this? I decided to create annotation, because it's the simplest way to have shared LB name for ingresses. 

I think I need to create some logging to point out users with the same hostname or with the same static IP in different ingresses to this feature.

File: pkg/controller/controller_test.go [hunk 1]
```diff
@@ -216,3 +225,209 @@ func TestEnsureMCIngress(t *testing.T) {
...
+		if pm.Name == pathMatcherName {
+			return pm, nil
+		}
+	}
+	return nil, fmt.Errorf("failed to find pathMatcher %+v", pathMatcherName)
+}
+
+func pathRuleExists(pathRules []*compute.PathRule, path string, service string) bool {
+	for _, pr := range pathRules {
+		//if pr.Service == service && slice.ContainsString(pr.Paths, path, nil) {
```
Comment (pondohva) [inline pkg/controller/controller_test.go]: can't find how I can translate backend name like "global/backendServices/k8s-be-30002--aaaaa" to k8s service name

**Comment by rramkumar1** [comment]
@pondohva Annotation seems like the only sensible way to do this. I'm taking a look now at your implementation

**Comment by pondohva** [comment]
@rramkumar1 
1. yes, sure. 
2. Thanks. I started to merge this in pkg/controller/controller.go:481, but I found that pass ingresslist instead of one ingress is not optimal decision. A lot of places in codebase expects 1 ingress in syncState. I'll think about this.

**Comment by pondohva** [comment]
@rramkumar1 could you please look again?

**Comment by pondohva** [comment]
BTW, I haven't tested it yet, just run go test.

**Comment by rramkumar1** [comment]
@pondohva I'm still having trouble understanding how your changes would even work. In particular, I don't see any sort of merging logic. Can you please explain your approach?

**Comment by pondohva** [comment]
@rramkumar1 I merged urlmap only in https://github.com/kubernetes/ingress-gce/pull/565/files#diff-3c862eb54a8e0e161b534e0c67e5379eR520.

this ingresses:

```
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: basic-ingress1
  annotations:
    kubernetes.io/ingress.class: "gce"
    kubernetes.io/ingress.loadbalancer-name: "abc"
spec:
  rules:
  - host: host.example.com
    http:
      paths:
      - backend:
          serviceName: web
          servicePort: 8080
        path: /web1
---
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: basic-ingress2
  annotations:
    kubernetes.io/ingress.class: "gce"
    kubernetes.io/ingress.loadbalancer-name: "abc"
spec:
  rules:
  - host: host.example.com
    http:
      paths:
      - backend:
          serviceName: web2
          servicePort: 8080
        path: /web2
```

result:
```
$ kubectl get ing
NAME             HOSTS              ADDRESS          PORTS     AGE
basic-ingress1   host.example.com   35.244.156.139   80        4m
basic-ingress2   host.example.com   35.244.156.139   80        4m
```

screenshot 2018-12-11 at 22 13 45

**Comment by pondohva** [comment]
@rramkumar1 I got your point. Added merge logic here https://github.com/kubernetes/ingress-gce/pull/565/files#diff-3c862eb54a8e0e161b534e0c67e5379eR377

**Comment by pondohva** [comment]
@rramkumar1 could you please review last changes? anything else to do?

**Comment by rramkumar1** [comment]
@pondohva Sorry I was on vacation and haven't had much time to look at this. 

Given that this is a big change I would like to spend some more time looking at it. I'll try to have some comments in a week or so.

**Comment by pondohva** [comment]
@rramkumar1 did you have some time to look at this PR? I'd like to implement this feature, but maybe I should rework smth?

**Comment by rramkumar1** [comment]
@pondohva Overall the PR seems to get the job done but my feeling is that the logic is going to be really flimsy and error prone. 

I'll leave this PR open but for the time being I don't think we will move forward with it. 

/lifecycle frozen

**Comment by mcfedr** [comment]
@rramkumar1 This seems like an interesting concept, do you have any comments what it would take to move forward with it?

**Comment by agadelshin** [comment]
thank you @bowei for update. I’ll rebase PR and will try to figure out why tests failed. 

Also I’d like to finalize it so if you have any suggestions how I can improve it, I’d appreciate.

**Comment by agadelshin** [comment]
oh, well, it's almost 1.5 years later. I believe it will be safer and better for learning to rewrite it :)

**Comment by rramkumar1** [comment]
@agadelshin I think you will find interest in the new Gateway APIs which are more tailored to solve this problem. I don't think we will be able to merge any form of this PR to this repo

https://github.com/kubernetes-sigs/service-apis

**Comment by agadelshin** [comment]
@rramkumar1 got it! thank you, I'll look into Gateway APIs.

**Pull request closed without merging**